src/Analyzer/Passes/OptimizeRedundantFunctionsInOrderByPass.cpp

```diff
@@ -0,0 +1,120 @@
+#include <Analyzer/Passes/OptimizeRedundantFunctionsInOrderByPass.h>
+#include <Analyzer/ColumnNode.h>
+#include <Analyzer/FunctionNode.h>
+#include <Analyzer/InDepthQueryTreeVisitor.h>
+#include <Analyzer/QueryNode.h>
+#include <Analyzer/SortNode.h>
+#include <Functions/IFunction.h>
+
+namespace DB
+{
+
+namespace
+{
+
+class OptimizeRedundantFunctionsInOrderByVisitor : public InDepthQueryTreeVisitor<OptimizeRedundantFunctionsInOrderByVisitor>
+{
+
+    struct RedundancyVerdict
+    {
+        bool redundant = true;
+        bool done = false;
+    };
+
+    static constexpr RedundancyVerdict makeNonRedundant() noexcept { return { .redundant = false, .done = true }; }
+
+    std::unordered_set<String> existing_keys;
+
+    RedundancyVerdict isRedundantExpression(FunctionNode * function)
```

Comment:
Should not we just return true if expression is redundant, and false otherwise. Column is redundant if we already saw it before. Constant is redundant by definition. Function is redundant if all arguments redundand. Or this are wrong assumptions ? I did not check original AST implementation.